client/src/main/scala/client/BackPressuredWebSocketClient.scala

```diff
@@ -1,93 +1,103 @@
+              case _: Throwable => ()
+            }
         }
 
-        webSocket.onclose = (event: CloseEvent) => {
-          downstream.onComplete()
+        val downstream = Observer.toReactiveSubscriber(subscriber)
+
+        try {
+          webSocket.onopen = (event: Event) => {
+            downstream.onSubscribe(new Subscription {
+              def cancel(): Unit =
+                closeConnection()
+
+              def request(n: Long): Unit = {
+                webSocket.send(n.toString)
+              }
+            })
+          }
+
+          webSocket.onerror = (event: ErrorEvent) => {
+            downstream.onError(BackPressuredWebSocketClient.Exception(event.message))
+          }
+
+          webSocket.onclose = (event: CloseEvent) => {
+            downstream.onComplete()
+          }
+
+          webSocket.onmessage = (event: MessageEvent) => {
+            downstream.onNext(event.data.asInstanceOf[String])
+          }
+        } catch {
+          case NonFatal(ex) =>
+            downstream.onError(ex)
         }
 
-        webSocket.onmessage = (event: MessageEvent) => {
-          downstream.onNext(event.data.asInstanceOf[String])
-        }
-      }
-      catch {
-        case ex: Throwable =>
-          downstream.onError(ex)
+        Cancelable(closeConnection)
       }
     }
 
-  private def closeConnection(webSocket: WebSocket)(implicit s: Scheduler): Unit = {
-    if (webSocket != null && webSocket.readyState <= 1)
-      try webSocket.close() catch { case _: Throwable => () }
-  }
-
-  def onSubscribe(subscriber: Subscriber[String]): Unit = {
+  override def unsafeSubscribeFn(subscriber: Subscriber[String]): Cancelable = {
     import subscriber.scheduler
 
-    var webSocket: WebSocket = null
-    val channel = try {
-      Utils.log(s"Connecting to $url")
-      webSocket = new WebSocket(url)
-      createChannel(webSocket)
-    }
-    catch {
-      case ex: Throwable =>
-        Observable.error(ex)
-    }
-
-    val source = channel.timeout(5.seconds)
-      .doOnCanceled(closeConnection(webSocket))
-
-    source.onSubscribe(new Observer[String] {
-      def onNext(elem: String): Future[Ack] =
-        subscriber.onNext(elem)
+    val subscription: Cancelable =
+      createChannel().subscribe(new Observer[String] {
+        def onNext(elem: String): Future[Ack] =
+          subscriber.onNext(elem)
+
+        def onError(ex: Throwable): Unit = {
+          subscription.cancel()
+          scheduler.reportFailure(ex)
+          // retry connection in a couple of secs
+          self
+            .delaySubscription(3.seconds)
+            .unsafeSubscribeFn(subscriber)
+        }
 
-      def onError(ex: Throwable): Unit = {
-        closeConnection(webSocket)
-        scheduler.reportFailure(ex)
-        // retry connection in a couple of secs
-        self.delaySubscription(3.seconds)
-          .onSubscribe(subscriber)
-      }
+        def onComplete(): Unit = {
+          subscription.cancel()
```

Comment:
This is a "forward reference". The compiler should have already warned you. It's incorrect. The point is that you can end up with a `subscription.cancel()` before the `.subscribe()` above returns. If you don't believe me, test this code with an `Observable.now(1)` which completes immediately. You'll probably get a null pointer exception.

1. Forward references to `Cancelable` objects are created by means of `SingleAssignmentCancelable` - that's its purpose
2. You don't need to cancel the subscription in `onComplete` unless you're doing something terribly wrong, because the source (e.g. `createChannel()`) should already do any required resource cleaning for you ;-)

In other words, delete this `subscription.cancel()`.